daffodil-lib/src/main/scala/org/apache/daffodil/xml/QNameBase.scala

```diff
@@ -183,27 +183,9 @@ object QName {
       else if (targetNamespace == NoNamespace) None
       else if (targetNamespace == UnspecifiedNamespace) None
       else {
-        val prefixes = NS.allPrefixes(targetNamespace, scope)
-        prefixes.length match {
-          case 0 => None
-          case 1 => Option(prefixes.head)
-          case _ => {
-            //
-            // suppose we have xmlns="..." xmlns:ex="..." xmlns:tns="..."
-            //
-            // We want to prefer ex as the prefix in this case.
-            // So we take the shortest prefix that isn't empty string.
-            //
-            val notNullPrefixes = prefixes.filter(_ ne null) // remove null prefix. There must be a non-empty one.
-            notNullPrefixes.foreach { p => Assert.invariant(p.length > 0) }
-            val pairs = notNullPrefixes.map { _.length } zip notNullPrefixes
-            val first = pairs.head // there has to be 1 at least. But there might be only 1
-            val minLengthPair = pairs.foldLeft(first) { case (x @ (xlen, _), y @ (ylen, _)) => if (xlen <= ylen) x else y }
-            val (n, shortest) = minLengthPair
-            Assert.invariant(n > 0)
-            Some(shortest)
-          }
-        }
+        val prefix = scope.getPrefix(targetNamespace)
```

Comment:
Why this change? The idea was to be deterministic rather than the prefix being just whatever scope.getPrefix happens to give us when there are multiple alternatives.

Reply:
So we found that it was this implemention was at odds with getPrefix. namedQName.prefix and thisElementNamespacePrefix were returning separate things. For the particular example I was looking at, I *think* the former was returning ex and the latter tns. @stevedlawrence  and I discussed it and we found it was more accurate to return the getPrefix implementation which is based on the order of the namespaces are they defined in the schema (which ends up being null, in the cases we have xmlns="...")

Reply:
So this returns the first defined prefix for the namespace? If so that's a useful thing that people can depend on which is to define the prefix they "like" as the canonical one first, and extra prefixes like tns that are really just for local schema-author convenience later in the list.

Reply:
Correct!